articles/remapping.md

```diff
@@ -0,0 +1,119 @@
+
+**A user can remap the first namespace token without replacing all namespace tokens with the same name.**
+
+*Example*
+Node uses names `/big/cat`, `/big/red/dog`, `/red/big/cat`.
+It must be possible to make a rule that replaces "big" with "small" such that the final names are `/small/cat`, `/small/red/dog`, `/red/big/cat`
+
+**A user can replace all matching namespace tokens without changing the basename.**
+
+*Example*
+Node uses names `/big/cat`, `/red/big/dog`, `/dog/big`.
+It must be possible to make a rule that replaces "big" with "small" such that the final names are `/small/cat`, `/red/small/dog`, `/dog/big`
+
+**A user can replace multiple matching namespace tokens only if whole namespaces are matched.**
+
+*Example*
+Node uses names `/big/red/dog`, `/very_big/red/dog`.
+It must be possible to make a rule that replaces "big/red" with "small" such that the final names are `/small/dog`, `/very_big/red/dog`
+
+##### 2. Basename Replacement
+**A user can replace a full basename token with another basename token**
+
+*Example*
+Node uses names `/big/cat`, `/big/cat/paws`.
+It must be possible to make a rule that replaces "cat" with "dog" such that the final names are `/big/dog`, `/big/cat/paws`
+
+##### 3. Exact Replacement
+**A user can replace a full name with another name**
+
+*Example*
+Node uses names `/big/cat`, `/big/cat/paws`.
+It must be possible to make a rule that replaces "/big/cat" with "/bear" such that the final names are `/bear`, `/big/cat/paws`
+
+##### 4. Arbitrary Substring Replacement ***(maybe)***
+**A user can replace any substring with another string**
+
+*Examples*
+Node uses names `/big/red/dog`, `/very_big/red/dog`.
+It must be possible to make a rule that replaces "big/red" with "small" such that the final names are `/small/dog`, `/very_small/dog`
+
+Node uses names `/big/dog`, `/dog/big`.
+It must be possible to make a rule that replaces "dog" with "cat" such that the final names are `/big/cat`, `/cat/big`
+
+Node uses names `/big/red/dog`, `/small/cat`.
+Node is run in namespace `ns`.
+It must be possible to make a rule that replaces "/" with "_" such that the final names are `/ns/_big_red_dog`, `/ns/_small_cat`
+
+#### Static Remapping
+Static remapping is the ability to provide a node with remap rules prior to launching it.
+A node stores all remapping rules for the duration of it's life.
+All names have remapping rules applied before they are used.
+This means the DDS interface, and maybe all of `rmw`, is ignorant of remapping.
+
+In ROS 1 remap rules are passed into the node at startup.
+`roscpp` received remap rules as arguments to `ros::init()`.
+`rospy` loaded remap rules from `sys.argv` during `rospy.init_node()`.
+
+In ROS 2 this code can be shared between client implementations by doing it in `rcl`.
+[rcl_init()](http://docs.ros2.org/beta1/api/rcl/rcl_8h.html#a6abc5188c50b3a4a5fc41b357b49d6e5) already requires command line arguments to be passed to it.
```

Comment:
In short, yes. In ROS 1, initialization (normally once per process) and node creation are conflated into `ros::init()`. In ROS 2, initialization is separate from node creation, and so we have to separate the responsibilities of `ros::init()` accordingly. For this case, processing command line arguments only really makes sense to do once per process, so maybe that should be inherent in the `rcl_init` function. So given the command line arguments as input, you should get back out modified command line arguments and some remappings. Then the remappings should be given somehow to each node at creation time. One way to do this is by passing them in as options to `rcl_node_init`

That all being said, I want to second the idea that we avoid any specifics like this, and instead speak in language neutral terms. For example, I'd say something like:

> Arguments passed on the command line can be converted into remapping rules, which can be shared with nodes and other ROS entities that the remapping rules might affect.

This implies what is possible without going to the point of calling out how it's done in C versus C++ or Python and without deciding if they should be implicitly shared or explicitly passed and to which functions. The former is more important information for the reader to gain a good conceptual understanding anyways. The discussion of implementation details may just add confusion at this point.